core/providers/java/java.go

```diff
@@ -0,0 +1,72 @@
+package java
+
+import (
+	"fmt"
+
+	"github.com/railwayapp/railpack/core/generate"
+	"github.com/railwayapp/railpack/core/plan"
+)
+
+type JavaProvider struct{}
+
+func (p *JavaProvider) Name() string {
+	return "java"
+}
+
+func (p *JavaProvider) Detect(ctx *generate.GenerateContext) (bool, error) {
+	return ctx.App.HasMatch("pom.{xml,atom,clj,groovy,rb,scala,yaml,yml}") || ctx.App.HasMatch("gradlew"), nil
+}
+
+func (p *JavaProvider) Initialize(ctx *generate.GenerateContext) error {
+	return nil
+}
+
+func (p *JavaProvider) StartCommandHelp() string {
+	return ""
+}
+
+func (p *JavaProvider) Plan(ctx *generate.GenerateContext) error {
+	p.installJDK(ctx)
+
+	build := ctx.NewCommandStep("build")
+	build.AddCommand(plan.NewCopyCommand("."))
+	build.Inputs = []plan.Input{plan.NewStepInput(ctx.GetMiseStepBuilder().Name())}
+
+	if p.usesGradle(ctx) {
+		p.installGradle(ctx)
+
+		if ctx.App.HasMatch("gradlew") && !ctx.App.IsFileExecutable("gradlew") {
+			build.AddCommand(plan.NewExecCommand("chmod +x gradlew"))
+		}
+
+		build.AddCommand(plan.NewExecCommand("./gradlew clean build -x check -x test"))
+		build.AddCache(p.gradleCache(ctx))
+	} else {
+		ctx.GetMiseStepBuilder().Default("maven", "latest")
+
+		if ctx.App.HasMatch("mvnw") && !ctx.App.IsFileExecutable("mvnw") {
+			build.AddCommand(plan.NewExecCommand("chmod +x mvnw"))
+		}
+
+		build.AddCommand(plan.NewExecCommand(fmt.Sprintf("%s -DoutputFile=target/mvn-dependency-list.log -B -DskipTests clean dependency:list install", p.getMavenExe(ctx))))
+	}
+
+	ctx.Deploy.Inputs = []plan.Input{
+		plan.NewStepInput(build.Name()),
```

Comment:
With Railpack, there is a difference between the build and runtime images. The build is heavy and includes a lot of system dependencies for compiling packages (like build-essential). The runtime image is much smaller.

The deploy inputs should be based on the runtime image and just copy over things from the build (like the /app directory and mise executables).

See [this](https://github.com/railwayapp/railpack/blob/main/core/providers/python/python.go#L72) as an example.